File: Data Structures & Algorithms/paint-house/submission-1.cpp
class Solution {
public:
    int minCost(vector<vector<int>>& costs) {
        int n = costs.size();
        vector<vector<int>> dp(n, vector<int>(3, 0));

        for (int i = 0; i < 3; ++i) {
            dp[n - 1][i] = costs[n - 1][i];
        }

        for (int r = n - 2; r >= 0; --r) {
            for (int c = 0; c < 3; ++c) {
                dp[r][c] = costs[r][c] + min(dp[r + 1][(c + 1) % 3], dp[r + 1][(c + 2) % 3]);
            }
        }

        return min(min(dp[0][0], dp[0][1]), dp[0][2]);
    }
};

File: Data Structures & Algorithms/min-cost-to-connect-points/submission-1.py
class Solution:
    def minCostConnectPoints(self, points: List[List[int]]) -> int:
        n, node = len(points), 0
        dist = [100000000] * n
        visit = [False] * n
        edges, res = 0, 0

        while edges < n - 1:
            visit[node] = True
            nextNode = -1

            for i in range(n):
                if visit[i]:
                    continue
                
                currDist = (abs(points[node][0] - points[i][0]) + 
                            abs(points[node][1] - points[i][1]))
                dist[i] = min(dist[i], currDist)
                if nextNode == -1 or dist[i] < dist[nextNode]:
                    nextNode = i

            res += dist[nextNode]
            edges += 1
            node = nextNode
        
        return res
